wp-lazy-loading.php

```diff
@@ -118,57 +120,115 @@ function wp_lazy_loading_enabled( $tag_name, $context ) {
-				 *
-				 * @since (TBD)
-				 *
-				 * @param string $default The filtered value, defaults to `lazy`.
-				 * @param string $tag_html The tag's HTML.
-				 * @param string $content The HTML containing the image tag.
-				 * @param string $context Optional. Additional context. Defaults to `current_filter()`.
-				 */
-				$value = apply_filters( 'wp_set_image_loading_attr', 'lazy', $tag_html, $content, $context );
+	$selected_images = array();
+	$attachment_ids  = array();
 
-				if ( $value ) {
-					if ( ! in_array( $value, array( 'lazy', 'eager' ), true ) ) {
-						$value = 'lazy';
-					}
+	foreach ( $matches[0] as $image ) {
+		if ( preg_match( '/wp-image-([0-9]+)/i', $image, $class_id ) ) {
+			$attachment_id = absint( $class_id[1] );
 
-					return str_replace( '<img', '<img loading="' . $value . '"', $tag_html );
-				}
+			if ( $attachment_id ) {
+				/*
+				 * If exactly the same image tag is used more than once, overwrite it.
+				 * All identical tags will be replaced later with 'str_replace()'.
+				 */
+				$selected_images[ $image ] = $attachment_id;
+
+				// Overwrite the ID when the same image is included more than once.
+				$attachment_ids[ $attachment_id ] = true;
 			}
+		}
+	}
+
+	if ( count( $attachment_ids ) > 1 ) {
+		/*
+		 * Warm the object cache with post and meta information for all found
+		 * images to avoid making individual database calls.
+		 */
+		_prime_post_caches( array_keys( $attachment_ids ), false, true );
+	}
+
+	$add_srcset_sizes = 'the_content' === $context;
+	$add_loading      = wp_lazy_loading_enabled( 'img', $context );
+
+	foreach ( $selected_images as $image => $attachment_id ) {
+		$image_meta = wp_get_attachment_metadata( $attachment_id );
+
+		$filtered_image = $image;
+
+		// Add 'srcset' and 'sizes' attributes if applicable.
+		if ( $add_srcset_sizes && false === strpos( $filtered_image, ' srcset=' ) ) {
+			$filtered_image = wp_image_add_srcset_and_sizes( $filtered_image, $image_meta, $attachment_id );
+		}
+
+		// Add 'loading' attribute if applicable.
+		if ( $add_loading && false === strpos( $filtered_image, ' loading=' ) ) {
+			$filtered_image = wp_image_add_loading( $filtered_image, $image_meta, $attachment_id, $content, $context );
+		}
+
+		$content = str_replace( $image, $filtered_image, $content );
+	}
+
+	return $content;
+}
+
+/**
+ * Adds a 'loading' attribute to an existing 'img' element.
+ *
+ * @since (TBD)
+ *
+ * @param string $image         An HTML 'img' element to be filtered.
+ * @param array  $image_meta    The image meta data as returned by 'wp_get_attachment_metadata()'.
+ * @param int    $attachment_id Image attachment ID.
+ * @param string $content       The HTML content that the 'img' element is part of.
+ * @param string $context       Additional context to pass to the filters.
+ * @return string Converted 'img' element with 'loading' attribute added.
+ */
+function wp_image_add_loading( $image, $image_meta, $attachment_id, $content, $context ) {
```

Comment:
Looking at the patch as a whole, not sure it makes sense for this to be a separate function. Also the name doesn't sound that well, perhaps. If it is to stay, perhaps something like 
`wp_image_add_loading_attribute( $image_tag, ... )`
to make it clear it is only for `<img>` HTML tags, etc.

Reply:
I think it makes sense for this to be separate, just like the existing function for srcset and sizes is separate. Regarding the name, I followed the convention set in that existing function too.

Reply:
> Regarding the name, I followed the convention set in that existing function too.

wouldn't call it a convention, it's rather a self-documenting thing: describe what the function does in 4-5 words or less. The `add srcset (and) sizes` sounds pretty specific, `add loading` is quite general though. Perhaps `add loading attr`? At least it suggests that there's an HTML attribute involved.

Reply:
Well, `srcset` and `sizes` are attributes just like `loading` and the term "attribute" is excluded there. But I don't have a strong opinion on this.